src/rust/engine/src/fs.rs

```diff
@@ -0,0 +1,836 @@
+pub enum PathGlob {
+  Root,
+  Wildcard {
+    canonical_dir: Dir,
+    symbolic_path: PathBuf,
+    wildcard: Pattern,
+  },
+  DirWildcard {
+    canonical_dir: Dir,
+    symbolic_path: PathBuf,
+    wildcard: Pattern,
+    remainder: Vec<Pattern>,
+  },
+}
+
+impl PathGlob {
+  pub fn root_stat() -> PathStat {
+    PathStat::dir(PathBuf::new(), Dir(PathBuf::new()))
+  }
+
+  fn wildcard(canonical_dir: Dir, symbolic_path: PathBuf, wildcard: Pattern) -> PathGlob {
+    PathGlob::Wildcard {
+      canonical_dir: canonical_dir,
+      symbolic_path: symbolic_path,
+      wildcard: wildcard,
+    }
+  }
+
+  fn dir_wildcard(canonical_dir: Dir, symbolic_path: PathBuf, wildcard: Pattern, remainder: Vec<Pattern>) -> PathGlob {
+    PathGlob::DirWildcard {
+      canonical_dir: canonical_dir,
+      symbolic_path: symbolic_path,
+      wildcard: wildcard,
+      remainder: remainder,
+    }
+  }
+
+  pub fn create(filespecs: &[String]) -> Result<Vec<PathGlob>, String> {
+    let mut path_globs = Vec::new();
+    for filespec in filespecs {
+      let canonical_dir = Dir(PathBuf::new());
+      let symbolic_path = PathBuf::new();
+      path_globs.extend(PathGlob::parse(canonical_dir, symbolic_path, filespec)?);
+    }
+    Ok(path_globs)
+  }
+
+  /**
+   * Given a filespec String relative to a canonical Dir and path, split it into path components
+   * while eliminating consecutive '**'s (to avoid repetitive traversing), and parse it to a
+   * series of PathGlob objects.
+   */
+  fn parse(canonical_dir: Dir, symbolic_path: PathBuf, filespec: &str) -> Result<Vec<PathGlob>, String> {
+    let mut parts = Vec::new();
+    let mut prev_was_doublestar = false;
+    for component in Path::new(filespec).components() {
+      let part =
+        match component {
+          Component::Prefix(..) | Component::RootDir =>
+            return Err(format!("Absolute paths not supported: {:?}", filespec)),
+          Component::CurDir =>
+            continue,
+          c => c.as_os_str(),
+        };
+
+      // Ignore repeated doublestar instances.
+      let cur_is_doublestar = *DOUBLE_STAR == part;
+      if prev_was_doublestar && cur_is_doublestar {
+        continue;
+      }
+      prev_was_doublestar = cur_is_doublestar;
+
+      // NB: Because the filespec is a String input, calls to `to_str_lossy` are not lossy; the
+      // use of `Path` is strictly for os-independent Path parsing.
+      parts.push(
+        Pattern::new(&part.to_string_lossy())
+          .map_err(|e| format!("Could not decode {:?} as UTF8: {:?}", filespec, e))?
+      );
+    }
+
+    PathGlob::parse_globs(canonical_dir, symbolic_path, &parts)
+  }
+
+  /**
+   * Given a filespec as Patterns, create a series of PathGlob objects.
+   */
+  fn parse_globs(
+    canonical_dir: Dir,
+    symbolic_path: PathBuf,
+    parts: &[Pattern]
+  ) -> Result<Vec<PathGlob>, String> {
+    if canonical_dir.0.as_os_str() == "." && parts.len() == 1 && *CURRENT_DIR == parts[0].as_str() {
```

Comment:
Indeed!... I think this is a holdout from when we had separate matching syntax for directories that looked for the `.`. It's now impossible to select the root, and I'm ok with that =P